src/lang.rs

```diff
@@ -383,1255 +791,1737 @@ fn is_keyword(c: char) -> bool {
+    use annotate_snippets::snippet::Snippet;
+    use expect_test::{expect, Expect};
+    use std::cell::RefCell;
+
+    use super::*;
+
+    fn check(actual: (Query, &Vec<String>), expect: Expect) {
+        let actual_errors = actual.1.join("\n");
+        let actual_combined = format!("{:#?}\n{}", actual.0, actual_errors);
+        expect.assert_eq(&actual_combined);
     }
 
-    #[test]
-    fn parse_expr_ident() {
-        expect!(expr, "foo", Expr::column("foo"));
-        expect!(
-            expr,
-            "foo.bar.baz[-10]",
-            Expr::Column {
-                head: DataAccessAtom::Key("foo".to_owned()),
-                rest: vec![
-                    DataAccessAtom::Key("bar".to_owned()),
-                    DataAccessAtom::Key("baz".to_owned()),
-                    DataAccessAtom::Index(-10)
-                ]
+    fn check_query(query_in: &str, expect: Expect) {
+        let mut errors = vec![];
+        let parsed = {
+            let qc = QueryContainer::new(
+                query_in.to_string(),
+                Box::new(VecErrorReporter::new(&mut errors)),
+            );
+            let r = query(&qc);
+
+            if r.is_ok() {
+                r.unwrap()
+            } else {
+                Query {
+                    search: Search::And(vec![]),
+                    operators: vec![],
+                }
             }
-        )
+        };
+
+        check((parsed, &errors), expect);
     }
 
-    #[test]
-    fn parse_expr_precedence() {
-        expect!(
-            expr,
-            "2 * 3 + 4 * 5 < 4 / 2 + 1",
-            Expr::Binary {
-                left: Box::new(Expr::Binary {
-                    left: Box::new(Expr::Binary {
-                        left: Box::new(Expr::Value(data::Value::Int(2))),
-                        op: BinaryOp::Arithmetic(ArithmeticOp::Multiply),
-                        right: Box::new(Expr::Value(data::Value::Int(3))),
-                    }),
-                    op: BinaryOp::Arithmetic(ArithmeticOp::Add),
-                    right: Box::new(Expr::Binary {
-                        left: Box::new(Expr::Value(data::Value::Int(4))),
-                        op: BinaryOp::Arithmetic(ArithmeticOp::Multiply),
-                        right: Box::new(Expr::Value(data::Value::Int(5))),
-                    }),
-                }),
-                op: BinaryOp::Comparison(ComparisonOp::Lt),
-                right: Box::new(Expr::Binary {
-                    left: Box::new(Expr::Binary {
-                        left: Box::new(Expr::Value(data::Value::Int(4))),
-                        op: BinaryOp::Arithmetic(ArithmeticOp::Divide),
-                        right: Box::new(Expr::Value(data::Value::Int(2))),
-                    }),
-                    op: BinaryOp::Arithmetic(ArithmeticOp::Add),
-                    right: Box::new(Expr::Value(data::Value::Int(1))),
-                })
-            }
-        );
+    struct VecErrorReporter<'a> {
```

Comment:
if you make this `Arc<Mutex<Vec<String>>` we can drop this lifetime + simplify the lifetimes in the trait as well

Reply:
changed